classes/fetch_services.py

```diff
@@ -136,60 +148,78 @@ def request(self, url: str):
-            response["duration"] = mins * 60 + seconds
-
-        if not response.get("idealized"):
-            # some sites will return a display name rather than unique
-            # usernames or they might return extra things in their
-            # response properties, so this is here to make adjustments
-            # for an ideal response, which will skip this step if loaded from the cache
-            match response["webpage_url_domain"]:
-                case "twitter.com":
-                    response["channel"] = response.get("uploader_id")
-                    response["title"] = response.get("title")[
-                        response.get("title").find("-") + 2 :
-                    ]
-                case "tiktok.com":
-                    response["channel"] = response.get("uploader")
-                    # could be attached to the if condition, but would feel
-                    # unorganizd if future sites are added that might need to use this
-
-            # bilibili.com and newgrounds.com sites don't seem to return information for the `channel`
-            # response property, so use `uploader` instead.
-            if response.get("channel") is None:
-                response["channel"] = response.get("uploader")
-
-            response["idealized"] = True
-
-        video_data = {
-            "title": response.get("title"),
-            "uploader": response.get("channel"),
+        return {
+            "title": video_data.get("title"),
+            "uploader": video_data.get("uploader"),
             "upload_date": upload_date,
-            "duration": response.get("duration"),
+            "duration": video_data.get("duration"),
         }
 
-        return video_data
+    # Some urls might have specific issues that should
+    # be handled here before they can be properly processed
+    # If yt-dlp gets any updates that resolve any of these issues
+    # then the respective case should be updated accordingly
+    def preprocess(self, url: str) -> dict | None:
+        pattern = r'https?://(www\.)?([^/#?\.]+)'
+        match = re.search(pattern, url)
+        site = match.group(2)
+        changes = {}
+
+        match site:
+            case "x":
+                url = "https://twitter" + url[url.find("://") + 4 :]
+                changes = self.preprocess(url)
+                changes["url"] = url
+            
+            case "twitter":
+                changes["channel"] = "uploader_id"
+                changes["title"] = lambda vid_data: f"X post by {vid_data.get("uploader_id")} ({self.hash_str(vid_data.get("title"))})"
+                if url[0 : url.rfind("/")].endswith("/video") and int(url[url.rfind("/") + 1:]) != 1:
+                    err("This post has several videos and the fetched duration is innacurate. So it has been ignored")
+                    changes["duration"] = None
+
+            case "newgrounds":
+                changes["channel"] = "uploader"
+                err("Missing Duration")
```

Comment:
Include contextual information in this error message to indicate what is causing the problem.